csrc/ir/internal_nodes.h

```diff
@@ -2298,4 +2298,150 @@ class SdpaFwdOp : public Expr {
       const std::vector<PolymorphicValue>& inputs) const override;
 };
 
+/*
+SDPA bwd node with same functionality at::_scaled_dot_product_flash_attention
+grad_query = [N, H, L, E]
+grad_key = [N, H, S, E]
+grad_value = [N, H, S, Ev]
+
+grad_output = [N, H, L, Ev]
+query = [N, H, L, E]
+key = [N, H, S, E]
+value = [N, H, S, Ev]
+output = [N, H, L, Ev]
+logsumexp = [N, H, L]
+cum_seq_q = [N + 1,] (undefined tensor for non-nested tensor)
+cum_seq_k = [N + 1,] (undefined tensor for non-nested tensor)
+query_seq_len = scalar(int)
+key_seq_len = scalar(int)
+dropout_p = scalar(double)
+is_causal = scalar(bool)
+philox_seed = scalar CPU tensor
+philox_offset = scalar CPU tensor
+scale = scalar(double)
+
+N = number of sequences / batch size
+H = num of heads
+L = query sequence length / target sequence length
+S = key/value sequence length / src sequence length
+E = query/key embd dimension
+Ev = value embd dimension
+
+For flash attention, E = Ev
+*/
+
+class SdpaBwdOp : public Expr {
+ public:
+  using Expr::Expr;
+
+  SdpaBwdOp(
+      IrBuilderPasskey,
+      TensorView* grad_query,
+      TensorView* grad_key,
+      TensorView* grad_value,
+      TensorView* grad_output,
+      TensorView* query,
+      TensorView* key,
+      TensorView* value,
+      TensorView* output,
+      TensorView* log_sumexp,
+      TensorView* cum_seq_q,
+      TensorView* cum_seq_k,
```

Comment:
Will these be `nullptr`? If so then we might segfault since we often assume `Expr::inputs()` returns non-null `Val*` pointers. One such place is in `toString()` below.

Reply:
I think we are getting empty tensors in those case. So it still wouldn't be nullptr?

Reply:
Ah ok, so we actually are getting tensors they just have size [0] or something?

Reply:
It's more explicit than that. On the cpp side, you can query `tensor.defined()`.

My rusty understanding is that, when a python API gets a None for a tensor input, you'll just have an undefined tensor. Which distinguishes it from actually receiving an empty tensor as input to the argument.

Reply:
Yes, it is an undefined tensor. It does not have a device, but it is not a `nullptr`.

Reply:
Interesting! Thanks for teaching me about this. Should we be checking `tensor.defined()` for inputs to `FusionExecutor::runFusion`? I suppose since their device is missing we would get an error about that if we were passed an undefined tensor in a case where we're not just evaluating with ATen.

Reply:
Should this be done at node level since codegen should not be dealing with undefined tensors anyway? That is, filtering out the undefined tensors or deciding if it needs to get replaced with a tensor of zeros (https://github.com/pytorch/pytorch/issues/41359) based on what is required for kernel generation. Right now, we don't have any concrete examples though for non-ATen-only nodes.

I am checking for `tensor.defined()` in `getCommonCUDADevice` and `validateValWithConcreteValue` which are called within `compileFusion` and `runFusion` (from within `kernelArgumentHolder`).

Reply:
I think since we don't have a way to handle undefined tensors at codegen yet (and maybe ever?) that it would be nice if we could error out if one is passed to a non-ATen evaluated segment. I suspect we'd just hit a less-informative error anyway though in such a case, and I'm not sure where we'd need to put that check. Since the approach you have here works for the ATen evaluated case, I think we're good.